### PR TITLE
Add support for Ceph Rados benchmarking.

### PR DESCRIPTION
**Pull request by ifed01**
Signed-off-by: Igor Fedotov <ifedotov@suse.com>

### CODE REVIEW

**Comment by axboe** [comment]
Looks fine to me. Can you add HOWTO and fio.1 additions as well to explain the new engine type?

**Comment by ifed01** [comment]
@axboe - thanks for looking into that. Will do next week.

File: engines/rados.c [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+	rados_completion_t completion;
+	rados_write_op_t write_op;
+	int io_seen;
+	int io_complete;
+};
+
+struct rados_data {
+	rados_t cluster;
+	rados_ioctx_t io_ctx;
+	char** objects;
```
Comment (axboe) [inline engines/rados.c]: Please follow the fio style, which would make this:

```
char **objects;
```
like you actually do for other members.

File: engines/rados.c [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+	size_t object_count;
+	struct io_u **aio_events;
+	bool connected;
+};
+
+/// fio configuration options read from the job file
```
Comment (axboe) [inline engines/rados.c]: No C++ style comments, use /**/

File: engines/rados.c [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+struct Options {
```
Comment (axboe) [inline engines/rados.c]: struct rados_options

File: engines/rados.c [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+	{
+		.name		= "pool",
+		.lname		= "pool name to use",
+		.type		= FIO_OPT_STR_STORE,
+		.help		= "Ceph pool name to benchmark against",
+		.off1		= offsetof(struct Options, pool_name),
+		.category	= FIO_OPT_C_ENGINE,
+		.group		= FIO_OPT_G_RBD,
+	},
+	{
```
Comment (axboe) [inline engines/rados.c]: space vs tabs confusion here

File: engines/rados.c [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+		r = rados_aio_create_completion(io_u,
+		on_complete_cb,
+		NULL /*cb_safe*/,
+		&fri->completion);
+		if (r < 0) {
+			log_err("rados_aio_create_completion failed.\n");
+			goto failed;
+		}
+
+		r = rados_aio_write(
```
Comment (axboe) [inline engines/rados.c]: Indentation. The fio standard is to use up 80 chars, then spill into the next line if you have more arguments.

File: engines/rados.c [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+			fri = u->engine_data;
+			if (fri->io_complete) {
+				fri->io_complete = 0;
+				rados->aio_events[events] = u;
+				events++;
+			}
+		}
+		if (events >= min)
+			break;
+		usleep(100);
```
Comment (axboe) [inline engines/rados.c]: You should not sleep unless ```t``` is set and indicates that the callers wants to wait for completions.

**Comment by ifed01** [comment]
@axboe - updated.
I decided to introduce an additional parameter 'heavy_load' to control whether straightforward rados poll or a wait on conditional variable are used. The rationale is a trade-off between CPU utilization and resulting throughput. In direct poll mode I'm getting ~10-15% better throughput at cost of 100% CPU utilization. Don't see any other ways to get such a performance without that CPU load. And not sure that high utilization is always good enough. Hence adding an option to be able to control this behavior.

**Comment by sitsofe** [comment]
@ifed01 re direct poll - is there any chance you can the option busy_poll ? There's precedent insofar as that's what the RBD engine calls it: (https://github.com/axboe/fio/blob/1a7bf7536e6cb4eda8354b527a639ca41f91d767/engines/rbd.c#L89 )...

**Comment by ifed01** [comment]
@sitsofe - renamed to busy_poll.

**Comment by axboe** [comment]
I've pulled this, but I'm hoping you'll also get the HOWTO updated for the engine and parameters.

**Comment by ifed01** [comment]
@axboe - thanks, will do.

**Pull request merged**